### ISSUE TITLE
Presubmit test-infra bazel test: reduce flakiness caused by image pull failures

### ISSUE THREAD

**Issue by chaodaiG** [kubernetes/test-infra#21027]
**What should be cleaned up or changed**:

It has been lately observed that presubmit bazel test has been pretty flaky on image pulling failures from code paths that are totally unrelated with the PR. Would be nice to reduce the flakiness by:

- Understand why bazel test needs to pull images, whether it's possible to avoid the pulling.
- If not possible, is there any chance to cache the image to make it less flaky?

**Provide any links for context**:

From PR: https://github.com/kubernetes/test-infra/pull/21026
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/21026/pull-test-infra-bazel/1364984456091275264
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/21026/pull-test-infra-bazel/1364983366922801152

From PR: https://github.com/kubernetes/test-infra/pull/20943
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/20943/pull-test-infra-bazel/1364685688871587840
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/20943/pull-test-infra-bazel/1364684891010109440
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/20943/pull-test-infra-bazel/1364687057552674816

**Comment by chaodaiG** [comment]
CC @alvaroaleman

**Comment by BenTheElder** [comment]
> Understand why bazel test needs to pull images, whether it's possible to avoid the pulling.

IIRC: `WORKSPACE` depencencies are resolved before doing any computation of the targets. Container images are workspace dependencies.

> If not possible, is there any chance to cache the image to make it less flaky?

Bazel doesn't support remote cache for this (`WORKSPACE`) using HTTP remote cache or RBE. It could be done by persisting the local `~/.cache/bazel`, though that can grow unbounded (normally a user might periodically delete it with e.g. `bazel clean --expunge`) and you'll need a solution for persisting it.

IIRC one of the issues is that these container pulls do not retry.

**Comment by alvaroaleman** [comment]
https://github.com/bazelbuild/rules_docker/issues/673 is the so far ignored bazel issue for this

**Comment by BenTheElder** [comment]
The test-infra bazel cache appears to be 16gb+ so sticking a copy in an image is probably not a great idea (as we've experimented with for .git), and uploading / downloading it will probably not behave super well as a single tarball in GCS either (as we've experimted with for go)

For pushing images it appears we added retries to calling bazel.

When testing test-infra we could call `bazel sync` or `bazel fetch ./...` in a retry loop before testing.

This would allow us to retry *only fetching dependencies* without retrying tests/builds.

I'll put up a PR.
